[JSC] Fix return in operationCopyOnWriteArrayIndexOfString

syg opened the pull request:
#### 773e875bad1457b0c41ce9b46671502354775406
<pre>
[JSC] Fix return in operationCopyOnWriteArrayIndexOfString
<a href="https://bugs.webkit.org/show_bug.cgi?id=294935">https://bugs.webkit.org/show_bug.cgi?id=294935</a>
<a href="https://rdar.apple.com/153522603">rdar://153522603</a>

Reviewed by Yusuke Suzuki, Keith Miller, and Ross Kirsling.

operationCopyOnWriteArrayIndexOfString was using RETURN_IF_EXCEPTION instead of
OPERATION_RETURN_IF_EXCEPTION, causing exceptions not propagate correctly since
they weren&apos;t being wrapped in ExceptionOperationResult. This meant code would
keep executing after an exception was thrown, triggering asserts downstream.

* JSTests/stress/regress-153522603.js: Added.
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):

Canonical link: <a href="https://commits.webkit.org/296649@main">https://commits.webkit.org/296649@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/335caba354a158b3482f746888a79c779115b8ec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109136 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19221 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114344 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59426 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29475 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37360 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82937 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112084 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23448 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98296 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63384 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22848 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16438 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/59011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101661 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92819 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16481 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117459 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/107694 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36180 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26749 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91952 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36552 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94559 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91758 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23376 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36672 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14426 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32034 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36077 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41581 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131964 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35770 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35766 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39107 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37456 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/JavaScriptCore/jit/OperationResult.h:
```diff
@@ -90,7 +90,7 @@ template<typename T>
 
 // This class exists so we can cast ExceptionOperationImplicitResult<From> to ExceptionOperationResult<To> implicitly while ExceptionOperationResult remains POD.
-// FIXME: we could use __atribute__((trivial_abi)) when we no longer support MSVC.
+// FIXME: we could use __attribute__((trivial_abi)) when we no longer support MSVC.
```
Hmm, probably should drop this line since it's unrelated. (On the other hand, this FIXME could totally be addressed, since we don't support MSVC anymore. :grin:)

syg replied:
Good to know, will remove.

rkirsling reviewed (approved):
Oho, a Shu patch at last!

Fix makes sense to me; just one comment.

kmiller68 commented on JSTests/stress/regress-153522603.js:
```diff
@@ -0,0 +1,6 @@
+//@ runDefault("--useFTLJIT=0", "--jitPolicyScale=0.1", "--watchdog=5000", "--watchdog-exception-ok")
```
Is there a reason to use a `5s` watchdog? Seems like we could get away with a much smaller number like 200ms or something.

syg replied:
Good call. I can repro locally easily enough in debug builds with 200ms. Note that I can't easily repro neither this test case or the original fuzz test in non-ASan release.

kmiller68 commented on Source/JavaScriptCore/jit/OperationResult.h:
```diff
@@ -90,7 +90,6 @@ template<typename T>
 using OperationReturnType = std::conditional_t<(std::is_same_v<T, void> || canMakeExceptionOperationResult<T>), ExceptionOperationResult<T>, T>;
 
 // This class exists so we can cast ExceptionOperationImplicitResult<From> to ExceptionOperationResult<To> implicitly while ExceptionOperationResult remains POD.
-// FIXME: we could use __atribute__((trivial_abi)) when we no longer support MSVC.
```
I think Ross meant to use the attribute and remove the class rather than delete the comment.

rkirsling replied:
Indeed; could be a separate PR if preferred though.

syg replied:
Yeah, since that involves removing this implicit coercion class, I'm reverting the removal and the typo fix to minimize this PR.

kmiller68 reviewed (approved):
r=me too.

syg committed:
[JSC] Fix return in operationCopyOnWriteArrayIndexOfString
https://bugs.webkit.org/show_bug.cgi?id=294935
rdar://153522603

Reviewed by Yusuke Suzuki, Keith Miller, and Ross Kirsling.

operationCopyOnWriteArrayIndexOfString was using RETURN_IF_EXCEPTION instead of
OPERATION_RETURN_IF_EXCEPTION, causing exceptions not propagate correctly since
they weren't being wrapped in ExceptionOperationResult. This meant code would
keep executing after an exception was thrown, triggering asserts downstream.

* JSTests/stress/regress-153522603.js: Added.
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):

Canonical link: https://commits.webkit.org/296649@main
